File: src/main/java/com/github/zametki/model/Zametka.java
package com.github.zametki.model;

import com.github.mjdbc.DbMapper;
import com.github.mjdbc.Mapper;
import org.jetbrains.annotations.NotNull;

import java.time.Instant;

public class Zametka extends Identifiable<ZametkaId> {

    public static final int MIN_CONTENT_LEN = 1;
    public static final int MAX_CONTENT_LEN = 50_000;

    @NotNull
    public UserId userId = UserId.INVALID_ID;

    @NotNull
    public Instant creationDate = Instant.MIN;

    @NotNull
    public String content = "";

    @NotNull
    public GroupId groupId = GroupId.UNDEFINED;

    @Mapper
    public static final DbMapper<Zametka> MAPPER = r -> {
        Zametka res = new Zametka();
        res.id = new ZametkaId(r.getInt("id"));
        res.userId = new UserId(r.getInt("user_id"));
        res.creationDate = r.getTimestamp("creation_date").toInstant();
        res.content = r.getString("content");
        res.groupId = new GroupId(r.getInt("group_id"));
        return res;
    };

    @Override
    public String toString() {
        return "Z[" + (id == null ? "?" : "" + id.value) + "]";
    }

}

File: src/main/java/com/github/zametki/model/Zametka.kt
package com.github.zametki.model

import com.github.mjdbc.DbMapper
import com.github.mjdbc.Mapper

import java.time.Instant

class Zametka : Identifiable<ZametkaId>() {

    var userId = UserId.INVALID_ID

    var creationDate = Instant.MIN!!

    var type = ZType.PLAIN_TEXT

    var content = ""

    var groupId = GroupId.ROOT

    override fun toString() = "Z[" + (if (id == null) "?" else "" + id!!.intValue) + "]"

    companion object {

        val MIN_CONTENT_LEN = 1
        val MAX_CONTENT_LEN = 65535

        @Mapper
        @JvmField
        val MAPPER = DbMapper {
            val res = Zametka()
            res.id = ZametkaId(it.getInt("id"))
            res.userId = UserId(it.getInt("user_id"))
            res.creationDate = it.getTimestamp("creation_date").toInstant()
            res.type = ZType.fromDbValue(it.getInt("type"))
            res.content = it.getString("content")
            res.groupId = GroupId(it.getInt("group_id"))
            res
        }
    }

}
